File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/MissileAccelerationCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator;

import com.github.aadvorak.artilleryonline.battle.calculations.MissileAccelerations;
import com.github.aadvorak.artilleryonline.battle.calculations.MissileCalculations;
import com.github.aadvorak.artilleryonline.battle.calculator.missile.CorrectingAccelerationCalculator;
import com.github.aadvorak.artilleryonline.battle.common.Acceleration;
import com.github.aadvorak.artilleryonline.battle.common.BodyAcceleration;
import com.github.aadvorak.artilleryonline.battle.model.BattleModel;

public class MissileAccelerationCalculator {

    public static BodyAcceleration calculate(MissileCalculations calculations, BattleModel battleModel) {
        var accelerations = new MissileAccelerations();

        accelerations.getGravity()
                .setY(-battleModel.getRoom().getSpecs().getGravityAcceleration());
        accelerations.setFriction(calculateFriction(calculations,
                battleModel.getRoom().getSpecs().getAirFrictionCoefficient()));
        accelerations.setPushing(calculatePushing(calculations));
        accelerations.setCorrecting(CorrectingAccelerationCalculator.calculate(calculations, battleModel));
        // todo returning
        return accelerations.sum();
    }

    private static Acceleration calculatePushing(MissileCalculations calculations) {
        var angle = calculations.getModel().getState().getPosition().getAngle();
        var pushingMagnitude = calculations.getModel().getSpecs().getPushingAcceleration();
        return new Acceleration()
                .setX(pushingMagnitude * Math.cos(angle))
                .setY(pushingMagnitude * Math.sin(angle));
    }

    private static BodyAcceleration calculateFriction(MissileCalculations calculations, double frictionCoefficient) {
        var velocity = calculations.getModel().getState().getVelocity();
        var positionAngle = calculations.getModel().getState().getPosition().getAngle();
        var velocityAngle = calculations.getModel().getState().getVelocity().getMovingVelocity().angle();
        var diffAngle = positionAngle - velocityAngle;
        var resultCoefficient = frictionCoefficient * (1 + Math.abs(Math.sin(diffAngle)));
        return new BodyAcceleration()
                .setX( - velocity.getX() * Math.abs(velocity.getX()) * resultCoefficient)
                .setY( - velocity.getY() * Math.abs(velocity.getY()) * resultCoefficient)
                .setAngle(- velocity.getAngle());
    }
}

File: frontend/playground/battle/calculator/missile-acceleration-calculator.ts
import type {BattleModel, MissileModel} from "~/playground/data/model";
import type {Acceleration, BodyAcceleration} from "~/playground/data/common";
import {VectorUtils} from "~/playground/utils/vector-utils";
import {
  CorrectingAccelerationCalculator
} from "~/playground/battle/calculator/missile/correcting-acceleration-calculator";

export const MissileAccelerationCalculator = {
  calculate(missileModel: MissileModel, battleModel: BattleModel): BodyAcceleration {
    const gravity = {
      x: 0.0,
      y: -battleModel.room.specs.gravityAcceleration
    }
    const pushing = this.calculatePushing(missileModel)
    const movingSum = VectorUtils.sumOf(gravity, pushing)

    const correcting = CorrectingAccelerationCalculator.calculate(missileModel, battleModel)

    const bodyAcceleration = {
      x: movingSum.x,
      y: movingSum.y,
      angle: correcting
    }
    const friction = this.calculateFriction(missileModel, battleModel.room.specs.airFrictionCoefficient)
    return VectorUtils.sumOfBody(bodyAcceleration, friction)
  },

  calculatePushing(missileModel: MissileModel): Acceleration {
    const angle = missileModel.state.position.angle
    const pushingMagnitude = missileModel.specs.pushingAcceleration
    return {
      x: pushingMagnitude * Math.cos(angle),
      y: pushingMagnitude * Math.sin(angle)
    }
  },

  calculateFriction(missileModel: MissileModel, frictionCoefficient: number): BodyAcceleration {
    const velocity = missileModel.state.velocity
    const positionAngle = missileModel.state.position.angle
    const velocityAngle = VectorUtils.getAngle(velocity)
    const diffAngle = positionAngle - velocityAngle
    const resultCoefficient = frictionCoefficient * 2 * (1 + 6 * Math.abs(Math.sin(diffAngle)))

    return {
      x: -velocity.x * Math.abs(velocity.x) * resultCoefficient,
      y: -velocity.y * Math.abs(velocity.y) * resultCoefficient,
      angle: -2.0 * velocity.angle
    }
  }
}
